File: trie/cpp/node.cpp
#include "node.h"
#include <iostream>

node_t::node_t() : is_end(false), is_word_end(false){}

void node_t::insert(std::string word) {
    node_t* curr = this;
    for(int i = 0; i < word.length(); i++){
        curr->is_end = false;
        char ch = word.at(i);
        if(curr->m1.find(ch) == curr->m1.end()){
            curr->m1.insert({ch, node_t()});
        }
        curr = &curr->m1.at(ch);
    }
    curr->is_end = true;
    curr->is_word_end = true;
}

void node_t::print(){
    print(this);
}

void node_t::print(const node_t* root) {
    if(root->is_end){
        return;
    }
    for(auto it = root->m1.begin(); it != root->m1.end(); it++){
        std::cout << it->first;
        if(!it->second.is_end){
            std::cout << "(";
            print(&it->second);
            std::cout << ")";
        }
    }
}

void node_t::print_words() {
    print_words(*this, "");
}

void node_t::print_words(const node_t& root, std::string str) {
    if(root.is_word_end){
        std::cout<< str << std::endl;
    }
    for(auto it = root.m1.begin(); it != root.m1.end(); it++){
        str.push_back(it->first);
        print_words(it->second, str);
        str.pop_back();
    }
}

bool node_t::get_strings_from_index(std::string str) {
    node_t* root = this;
    for(int i = 0; i < str.length(); i++){
        char ch = str.at(i);
        if(root->m1.find(ch) == root->m1.end()){
            return false;
        }
        root = &root->m1.at(ch);
    }
    print_words(*root, str);
    return true;
}

File: trie/java/Node.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.HashMap;
import java.util.Scanner;

public class Node{
    HashMap<Character, Node> map;
    boolean isEnd;
    boolean wordEnd;

    Node(){
        isEnd = false;
        wordEnd = false;
        map = new HashMap<>();
    }

    private void insert(String word){
        Node curr = this;
        for(int i = 0; i < word.length(); i++){
            curr.isEnd = false;
            char ch = word.charAt(i);
            if(!curr.map.containsKey(ch)){
                curr.put(ch);
            }
            curr = curr.get(ch);
        }
        curr.isEnd = true;
        curr.wordEnd = true;
    }

    private Node get(char ch){
        return this.map.get(ch);
    }

    private Node put(char ch){
        Node node = new Node();
        this.map.put(ch, node);
        return node;
    }

    private static void print(Node root){
        if(root.isEnd){
            return;
        }
        for(char ch : root.map.keySet()){
            System.out.print(ch);
            if(!root.get(ch).isEnd){
                System.out.print("(");
                print(root.get(ch));
                System.out.print(")");
            }
        }
    }

    private static void printWords(Node root, StringBuilder str){
        if(root.wordEnd){
            System.out.println(str.toString());
        }
        for(char ch : root.map.keySet()){
            str.append(ch);
            printWords(root.get(ch), str);
            str.deleteCharAt(str.length() - 1);
        }
    }

    private static void printWords(Node root){
        printWords(root, new StringBuilder());
    }

    private static boolean getStringsStartingWith(String str, Node root){
        for(char ch : str.toCharArray()){
            if(!root.map.containsKey(ch)){
                return false;
            }
            root = root.get(ch);
        }
        printWords(root, new StringBuilder(str));
        return true;
    }

    public static void main(String[] args) {
        Node trie = new Node();
        // try {
        //     // Scanner scan = new Scanner(new File("./text.txt"));
        //     // while(scan.hasNext()){
        //     //     trie.insert(scan.next().toLowerCase().trim());
        //     // }
        //     // scan.close();
        // } catch (FileNotFoundException e) {
        //     e.printStackTrace();
        // }
        trie.insert("cat");
        trie.insert("cast");
        print(trie);
        while(true){
            Scanner scan = new Scanner(System.in);
            getStringsStartingWith(scan.nextLine(), trie);
        }

    }
}